flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/ChangelogMapState.java

```diff
@@ -51,16 +63,28 @@ public UV get(UK key) throws Exception {
 
     @Override
     public void put(UK key, UV value) throws Exception {
+        if (getValueSerializer() instanceof MapSerializer) {
+            changeLogger.valueElementChanged(
+                    out -> {
+                        serializeKey(key, out);
+                        serializeValue(value, out);
+                    },
+                    getCurrentNamespace());
+        } else {
+            changeLogger.valueAdded(singletonMap(key, value), getCurrentNamespace());
```

Comment:
I wonder is it possible that the value serializer of chenagelogMap state is not a `MapSerializer`?

Reply:
I see some checks in both Heap and RocksDB backends that disallow other serializers.
But it seems an implementation detail and I'm not sure that we can depend on it.
WDYT?

Reply:
In general, I don't think it's an implementation detail. `MapStateDescriptor` only allows `MapSerializer` or `MapTypeInfo`, and the only serializer would be created in `MapTypeInfo` is `MapSerializer`: see code [here](https://github.com/apache/flink/blob/4d73310ce6152e4d6bdccca44b8e1f8ecec0e1b2/flink-core/src/main/java/org/apache/flink/api/java/typeutils/MapTypeInfo.java#L115).

Reply:
I'm not sure (as `MapStateDescriptor` is not the only possible `StateDescriptor`); but it simplifies the code :) I'll remove the additional logic, thanks